diff --git a/CHANGES b/CHANGES
index 65d9c15..3ac65ed 100644
--- a/CHANGES
+++ b/CHANGES
@@ -19,6 +19,9 @@ Daniel Stenberg (22 Mar 2010)
   that instead of simply returning without cleaning up, the code should have a
   goto that jumps to the spot right after the call to Curl_resolv().
 
+Kamil Dudka (19 Mar 2010)
+- Improved Curl_read() to not ignore the error returned from Curl_ssl_recv().
+
 Kamil Dudka (12 Nov 2009)
 - Kevin Baughman provided a fix preventing libcurl-NSS from crash on doubly
   closed NSPR descriptor. The issue was hard to find, reported several times
diff --git a/lib/sendf.c b/lib/sendf.c
index d28cb52..33110fe 100644
--- a/lib/sendf.c
+++ b/lib/sendf.c
@@ -535,9 +535,11 @@ int Curl_read(struct connectdata *conn, /* connection data */
   if(conn->ssl[num].state == ssl_connection_complete) {
     nread = Curl_ssl_recv(conn, num, buffertofill, bytesfromsocket);
 
-    if(nread == -1) {
+    if(nread == -1)
       return -1; /* -1 from Curl_ssl_recv() means EWOULDBLOCK */
-    }
+    else if(nread == -2)
+      /* -2 from Curl_ssl_recv() means a true error, not EWOULDBLOCK */
+      return CURLE_RECV_ERROR;
   }
   else if(Curl_ssh_enabled(conn, (PROT_SCP|PROT_SFTP))) {
     if(conn->protocol & PROT_SCP)
diff --git a/lib/sslgen.c b/lib/sslgen.c
index 17cb3ef..021733e 100644
--- a/lib/sslgen.c
+++ b/lib/sslgen.c
@@ -399,7 +399,7 @@ struct curl_slist *Curl_ssl_engines_list(struct SessionHandle *data)
   return curlssl_engines_list(data);
 }
 
-/* return number of sent (non-SSL) bytes */
+/* return number of sent (non-SSL) bytes; -1 on error */
 ssize_t Curl_ssl_send(struct connectdata *conn,
                       int sockindex,
                       const void *mem,
@@ -411,8 +411,8 @@ ssize_t Curl_ssl_send(struct connectdata *conn,
 /* return number of received (decrypted) bytes */
 
 /*
- * If the read would block (EWOULDBLOCK) we return -1. Otherwise we return
- * a regular CURLcode value.
+ * If the read would block (EWOULDBLOCK) we return -1. If an error occurs during
+ * the read, we return -2. Otherwise we return the count of bytes transfered.
  */
 ssize_t Curl_ssl_recv(struct connectdata *conn, /* connection data */
                       int sockindex,            /* socketindex */
@@ -425,9 +425,9 @@ ssize_t Curl_ssl_recv(struct connectdata *conn, /* connection data */
   nread = curlssl_recv(conn, sockindex, mem, len, &block);
   if(nread == -1) {
     if(!block)
-      return 0; /* this is a true error, not EWOULDBLOCK */
+      return -2; /* this is a true error, not EWOULDBLOCK */
     else
-      return -1;
+      return -1; /* EWOULDBLOCK */
   }
 
   return nread;
